packages/@lwc/template-compiler/src/codegen/static-element.ts

```diff
@@ -0,0 +1,217 @@
+    isIdReferencingAttribute,
+    isSvgUseHref,
+} from '../parser/attribute';
+import State from '../state';
+import { isCustomRendererHookRequired } from '../shared/renderer-hooks';
+
+// This set keeps track of static safe elements that have dynamic text in their direct children.
+const STATIC_ELEMENT_WITH_DYNAMIC_TEXT_SET = new WeakSet<StaticElement>();
+
+// This map keeps track of static safe elements to their transformed children.
+// The children are transformed so that contiguous text nodes are consolidated into arrays.
+const STATIC_ELEMENT_TO_DYNAMIC_TEXT_CHILDREN_CACHE = new WeakMap<
+    StaticElement,
+    (StaticChildNode | Text[])[]
+>();
+
+function isStaticNode(node: BaseElement, apiVersion: APIVersion): boolean {
+    let result = true;
+    const { name: nodeName, namespace = '', attributes, directives, properties } = node;
+
+    // SVG is excluded from static content optimization in older API versions due to issues with case sensitivity
+    // in CSS scope tokens. See https://github.com/salesforce/lwc/issues/3313
+    if (
+        !isAPIFeatureEnabled(APIFeature.LOWERCASE_SCOPE_TOKENS, apiVersion) &&
+        namespace !== HTML_NAMESPACE
+    ) {
+        return false;
+    }
+
+    // it is an element
+    result &&= isElement(node);
+
+    // all attrs are static-safe
+    // the criteria to determine safety can be found in computeAttrValue
+    result &&= attributes.every(({ name, value }) => {
+        const isStaticSafeLiteral =
+            isLiteral(value) &&
+            name !== 'slot' &&
+            // check for ScopedId
+            name !== 'id' &&
+            name !== 'spellcheck' && // spellcheck is specially handled by the vnodes.
+            !isIdReferencingAttribute(name) &&
+            // svg href needs sanitization.
+            !isSvgUseHref(nodeName, name, namespace) &&
+            // Check for ScopedFragId
+            !(
+                isAllowedFragOnlyUrlsXHTML(nodeName, name, namespace) &&
+                isFragmentOnlyUrl(value.value as string)
+            );
+        const isStaticSafeExpression =
+            isExpression(value) &&
+            name !== 'slot' &&
+            // TODO [#3624]: Revisit whether svgs can be included in static content optimization
+            // svg href needs sanitization.
+            !isSvgUseHref(nodeName, name, namespace);
+        return isStaticSafeLiteral || isStaticSafeExpression;
+    });
+
+    // all directives are static-safe
+    result &&= !directives.some((directive) => !STATIC_SAFE_DIRECTIVES.has(directive.name));
+
+    // all properties are static
+    result &&= properties.every((prop) => isLiteral(prop.value));
+
+    return result;
+}
+
+function collectStaticNodes(node: ChildNode, staticNodes: Set<ChildNode>, state: State) {
+    let childrenAreStaticSafe = true;
+    let nodeIsStaticSafe;
+
+    if (isText(node)) {
+        nodeIsStaticSafe = true;
+    } else if (isComment(node)) {
+        nodeIsStaticSafe = true;
```

Comment:
According to [the preserve-comments RFC](https://github.com/salesforce/lwc-rfcs/blob/master/text/0000-template-comments.md), we support `lwc:inner-html` for comments. So I'm not sure these are always static-safe.